rewrite-core/src/main/java/org/openrewrite/SourceFile.java

```diff
@@ -26,6 +27,18 @@
 import java.util.function.Predicate;
 
 public interface SourceFile extends Tree {
+
+    /**
+     * Does this source file represented as an LST, when printed, produce a byte-for-byte identical
+     * result to the original input source file?
+     *
+     * @param input The input source.
+     * @return <code>true</code> if the parse-to-print loop is idempotent, <code>false</code> otherwise.
+     */
+    default boolean printEqualsInput(Parser.Input input, ExecutionContext ctx) {
+        return printAll().equals(StringUtils.readFully(input.getSource(ctx)));
```

Comment:
Had thought to see if I could help out with the test failures in https://ge.openrewrite.org/s/crdmk2pb27noq/tests/overview?outcome=failed .
I'd locally expanded this to the following, such that I could set a conditional breakpoint on `!printed.equals(read)`.
```suggestion
        String printed = printAll();
        String read = StringUtils.readFully(input.getSource(ctx));
        return printed.equals(read);
```

Reply:
That then fails in 
https://github.com/openrewrite/rewrite/blob/dc9f04ae6dd44ac6c981fa63c5ee165d85c8f3ed/rewrite-java-test/src/test/java/org/openrewrite/java/JavaTemplateInstanceOfTest.java#L208-L220
with an excess `}}returnnull;` in the printed result.
```
tim@tim-xps-15-9520:~/.config/JetBrains/IntelliJIdea2023.1/scratches$ cat read 
class __P__ {  static native <T> T p();  static native <T> T[] arrp();  static native boolean booleanp();  static native byte bytep();  static native char charp();  static native double doublep();  static native int intp();  static native long longp();  static native short shortp();  static native float floatp();}class __M__ {  static native Object any(Object o);  static native Object any(java.util.function.Predicate<Boolean> o);  static native <T> Object anyT();}class T{Object m(Object o){if(true) {if (((Object)null) instanceof String s) {} else {__M__.any(
/*__TEMPLATE__*/s.length()/*__TEMPLATE_STOP__*/
)}}
return null;
}}
```
```
tim@tim-xps-15-9520:~/.config/JetBrains/IntelliJIdea2023.1/scratches$ cat printed 
class __P__ {  static native <T> T p();  static native <T> T[] arrp();  static native boolean booleanp();  static native byte bytep();  static native char charp();  static native double doublep();  static native int intp();  static native long longp();  static native short shortp();  static native float floatp();}class __M__ {  static native Object any(Object o);  static native Object any(java.util.function.Predicate<Boolean> o);  static native <T> Object anyT();}class T{Object m(Object o){if(true) {if (((Object)null) instanceof String s) {} else {__M__.any(
/*__TEMPLATE__*/s.length()/*__TEMPLATE_STOP__*/
)}}
return null;
}}returnnull;
}}
```

Reply:
Workaround for this preexisting issue in https://github.com/openrewrite/rewrite/pull/3459/commits/788040479ad0a28a628fab8bc405943078cea36f ; we can revisit at a more convenient time if needed.